Replace support

medranocalvo opened the pull request:
Support for replacing other window managers and being replaced by them.  Tested with FVWM2.

As a side-effect (and probably the greatest value of these changes) EXWM now cleans up after itself.

Needs chages to XELB: https://github.com/ch11ng/xelb/pull/19.

Please, review carefully.

ch11ng commented:
@medranocalvo Thanks! You've made a lot of PRs!  I'll try to finish the first round review by the end of this week.

medranocalvo commented:
No hurry :-).  Have a nice day.

tumashu commented:
do you have signed gnu paper?

medranocalvo commented:
@tumashu Yes I did.

medranocalvo commented on exwm-input.el:
```diff
@@ -967,14 +986,42 @@ where both ORIGINAL-KEY and SIMULATED-KEY are key sequences."
 
 (defun exwm-input--exit ()
```
I'll remove the cleanup code.

medranocalvo commented:
Dropped cleanup code.  Please, review again.

These are the changes from v1 (note I fixed a bug regarding exwm-exit):
```diff
diff --git c/exwm-input.el w/exwm-input.el
index 090e86b..5d0ca6c 100644
--- c/exwm-input.el
+++ w/exwm-input.el
@@ -429,13 +429,6 @@ ARGS are additional arguments to CALLBACK."
     (with-slots (window) evt
       (exwm-input--grab-global-prefix-keys window))))
 
-(defun exwm-input--global-prefix-key-window-list ()
-  "List of windows keys should be grabbed."
-  (slot-value (xcb:+request-unchecked+reply exwm--connection
-                  (make-instance 'xcb:QueryTree
-                                 :window exwm--root))
-              'children))
-
 (defun exwm-input--update-global-prefix-keys ()
   "Update `exwm-input--global-prefix-keys'."
   (when exwm--connection
@@ -445,7 +438,10 @@ ARGS are additional arguments to CALLBACK."
         (cl-pushnew (elt i 0) exwm-input--global-prefix-keys))
       (unless (equal original exwm-input--global-prefix-keys)
         (apply #'exwm-input--grab-global-prefix-keys
-               (exwm-input--global-prefix-key-window-list))))))
+               (slot-value (xcb:+request-unchecked+reply exwm--connection
+                               (make-instance 'xcb:QueryTree
+                                              :window exwm--root))
+                           'children))))))
 
 (defun exwm-input--grab-global-prefix-keys (&rest xwins)
   (let ((req (make-instance 'xcb:GrabKey
@@ -472,17 +468,6 @@ ARGS are additional arguments to CALLBACK."
           (xcb:+request exwm--connection req))))
     (xcb:flush exwm--connection)))
 
-(defun exwm-input--ungrab-global-prefix-keys (&rest xwins)
-  "Ungrab global prefix keys on windows XWINS."
-  (let ((ungrab-keys (make-instance 'xcb:UngrabKey
-                                    :key xcb:Grab:Any
-                                    :grab-window nil
-                                    :modifiers xcb:ModMask:Any)))
-    (dolist (xwin xwins)
-      (setf (slot-value ungrab-keys 'grab-window) xwin)
-        (if (xcb:+request-checked+request-check exwm--connection ungrab-keys)
-            (exwm--log "Failed to ungrab keys")))))
-
 (defun exwm-input--set-key (key command)
   (global-set-key key command)
   (cl-pushnew key exwm-input--global-keys))
@@ -986,42 +971,15 @@ where both ORIGINAL-KEY and SIMULATED-KEY are key sequences."
 
 (defun exwm-input--exit ()
   "Exit the input module."
-  ;; Ungrab keys
   (exwm-input--unset-simulation-keys)
-  (apply #'exwm-input--ungrab-global-prefix-keys
-         (exwm-input--global-prefix-key-window-list))
-  ;; Detach hooks.
   (remove-hook 'minibuffer-setup-hook #'exwm-input--on-minibuffer-setup)
   (remove-hook 'pre-command-hook #'exwm-input--on-pre-command)
   (remove-hook 'post-command-hook #'exwm-input--on-post-command)
   (remove-hook 'buffer-list-update-hook #'exwm-input--on-buffer-list-update)
-  ;; Detach event listeners.
-  (xcb:+remove-event exwm--connection 'xcb:PropertyNotify
-                     #'exwm-input--on-PropertyNotify)
-  (xcb:+remove-event exwm--connection 'xcb:CreateNotify
-                     #'exwm-input--on-CreateNotify)
-  (xcb:+remove-event exwm--connection 'xcb:KeyPress
-                     #'exwm-input--on-KeyPress)
-  (xcb:+remove-event exwm--connection 'xcb:ButtonPress
-                     #'exwm-input--on-ButtonPress)
-  (xcb:+remove-event exwm--connection 'xcb:ButtonRelease
-                     #'exwm-floating--stop-moveresize)
-  (xcb:+remove-event exwm--connection 'xcb:MotionNotify
-                     #'exwm-floating--do-moveresize)
-  (when mouse-autoselect-window
-    (xcb:+remove-event exwm--connection 'xcb:EnterNotify
-                       #'exwm-input--on-EnterNotify))
-  ;; Cancel timers.
   (when exwm-input--update-focus-defer-timer
     (cancel-timer exwm-input--update-focus-defer-timer))
   (when exwm-input--update-focus-timer
-    (cancel-timer exwm-input--update-focus-timer))
-  ;; Destroy windows
-  (when exwm-input--timestamp-window
-    (xcb:+request exwm--connection
-        (make-instance 'xcb:DestroyWindow
-                       :window exwm-input--timestamp-window)))
-  (setq exwm-input--timestamp-window nil))
+    (cancel-timer exwm-input--update-focus-timer)))
 
 
 
diff --git c/exwm-manage.el w/exwm-manage.el
index 9fa1cc7..b647819 100644
--- c/exwm-manage.el
+++ w/exwm-manage.el
@@ -622,14 +622,6 @@ border-width: %d; sibling: #x%x; stack-mode: %d"
   "Exit the manage module."
   (dolist (pair exwm--id-buffer-alist)
     (exwm-manage--unmanage-window (car pair) 'quit))
-  (xcb:+remove-event exwm--connection 'xcb:DestroyNotify
-                     #'exwm-manage--on-DestroyNotify)
-  (xcb:+remove-event exwm--connection 'xcb:UnmapNotify
-                     #'exwm-manage--on-UnmapNotify)
-  (xcb:+remove-event exwm--connection 'xcb:MapRequest
-                     #'exwm-manage--on-MapRequest)
-  (xcb:+remove-event exwm--connection 'xcb:ConfigureRequest
-                     #'exwm-manage--on-ConfigureRequest)
   (remove-hook 'after-make-frame-functions #'exwm-manage--add-frame)
   (remove-hook 'delete-frame-functions #'exwm-manage--remove-frame)
   (setq exwm-manage--_MOTIF_WM_HINTS nil))
diff --git c/exwm-randr.el w/exwm-randr.el
index 9a55d0b..c2860d4 100644
--- c/exwm-randr.el
+++ w/exwm-randr.el
@@ -217,16 +217,7 @@ the first one in result being the primary output."
 
 (defun exwm-randr--exit ()
   "Exit the RandR module."
-  (unless (or (not (xcb:got-extension-data-p exwm--connection 'xcb:randr))
-              (= 0 (slot-value (xcb:get-extension-data exwm--connection 'xcb:randr) 'present)))
-    (remove-hook 'exwm-workspace-list-change-hook #'exwm-randr--refresh)
-    (xcb:+remove-event exwm--connection 'xcb:randr:ScreenChangeNotify
-                       #'exwm-randr--on-ScreenChangeNotify)
-    (xcb:+request exwm--connection
-        (make-instance 'xcb:randr:SelectInput
-                       :window exwm--root
-                       :enable 0))
-    (xcb:flush exwm--connection)))
+  (remove-hook 'exwm-workspace-list-change-hook #'exwm-randr--refresh))
 
 (defun exwm-randr-enable ()
   "Enable RandR support for EXWM."
diff --git c/exwm-workspace.el w/exwm-workspace.el
index d17fb03..1d0d6b3 100644
--- c/exwm-workspace.el
+++ w/exwm-workspace.el
@@ -126,7 +126,6 @@ Please manually run the hook `exwm-workspace-list-change-hook' afterwards.")
 (defvar exwm-input--during-command)
 (defvar exwm-layout-show-all-buffers)
 (defvar exwm-manage--desktop)
-(declare-function exwm--exit "exwm.el")
 (declare-function exwm-input--on-buffer-list-update "exwm-input.el" ())
 (declare-function exwm-layout--hide "exwm-layout.el" (id))
 (declare-function exwm-layout--other-buffer-predicate "exwm-layout.el"
@@ -1547,7 +1546,7 @@ applied to all subsequently created X frames."
                                  ;; (client . nil)
                                  (fullscreen . nil)
                                  (buffer-predicate . nil))))
-  ;; Restore the 'client' frame parameter (before `exwm--exit').
+  ;; Restore the 'client' frame parameter (before `exwm-exit').
   (when exwm-workspace--client
     (dolist (f exwm-workspace--list)
       (set-frame-parameter f 'client exwm-workspace--client))
diff --git c/exwm.el w/exwm.el
index b8fff06..eaa68f4 100644
--- c/exwm.el
+++ w/exwm.el
@@ -685,21 +685,6 @@
                      :data [0 0]))
   (xcb:flush exwm--connection))
 
-(defun exwm--exit-icccm-ewmh ()
-  "Remove ICCCM/EWMH properties."
-  ;; Stop handling relevant client messages
-  (xcb:+remove-event exwm--connection 'xcb:ClientMessage #'exwm--on-ClientMessage)
-  ;; Stop handling PropertyNotify event
-  (xcb:+remove-event exwm--connection 'xcb:PropertyNotify #'exwm--on-PropertyNotify)
-  ;; Stop handling SelectionClear event
-  (xcb:+remove-event exwm--connection 'xcb:SelectionClear #'exwm--on-SelectionClear)
-  ;; Delete _NET_SUPPORTING_WM_CHECK's window
-  (when exwm--guide-window
-    (xcb:+request exwm--connection
-        (make-instance 'xcb:DestroyWindow :window exwm--guide-window))
-    (setq exwm--guide-window nil))
-  (xcb:flush exwm--connection))
-
 (defun exwm--wmsn-acquire (replace)
   "Acquire the WM_Sn selection.
 
@@ -775,14 +760,6 @@ manager.  If t, replace it, if nil, abort and if `ask'."
         (xcb:+request exwm--connection se))
       (setq exwm--wmsn-window new-owner))))
 
-(defun exwm--wmsn-release ()
-  "Release the window manager selection."
-  (when exwm--wmsn-window
-    (xcb:+request exwm--connection
-        (make-instance 'xcb:DestroyWindow
-                       :window exwm--wmsn-window))
-    (setq exwm--wmsn-window nil)))
-
 ;;;###autoload
 (defun exwm-init (&optional frame)
   "Initialize EXWM."
@@ -849,12 +826,10 @@ manager.  If t, replace it, if nil, abort and if `ask'."
   (exwm-workspace--exit)
   (exwm-floating--exit)
   (exwm-layout--exit)
-  (exwm--exit-icccm-ewmh)
   (xcb:+request-checked+request-check exwm--connection
       (make-instance 'xcb:ChangeWindowAttributes
                      :window exwm--root :value-mask xcb:CW:EventMask
                      :event-mask xcb:EventMask:NoEvent))
-  (exwm--wmsn-release)
   (when exwm--connection
     (xcb:flush exwm--connection)
     (xcb:disconnect exwm--connection))
@@ -961,7 +936,7 @@ manager.  If t, replace it, if nil, abort and if `ask'."
       (run-hooks 'kill-emacs-hook)
       (setq kill-emacs-hook nil))
     ;; Exit each module, destroying all resources created by this connection.
-    (exwm--exit)
+    (exwm-exit)
     ;; Set the return value.
     t))
```

medranocalvo commented:
Fixed leftover, rebased.

```diff
diff --git c/exwm.el w/exwm.el
index dfb0362..90b1fe4 100644
--- c/exwm.el
+++ w/exwm.el
@@ -119,7 +119,7 @@
 (defun exwm-restart ()
   "Restart EXWM."
   (interactive)
-  (when (exwm-workspace--confirm-kill-emacs "[EXWM] Restart? " 'no-check)
+  (when (exwm--confirm-kill-emacs "[EXWM] Restart? " 'no-check)
     (let* ((attr (process-attributes (emacs-pid)))
            (args (cdr (assq 'args attr)))
            (ppid (cdr (assq 'ppid attr)))
```

ch11ng commented:
Replacing EXWM with Openbox would fail due to Openbox receives SIGFPE.  I suspect there are some properties set to 0 by EXWM.  I'll fix that when merging this PR.

medranocalvo commented:
The issue is caused by these properties that EXWM leaves upon exit:
```
$ xprop -root | grep '^_NET_'
[...]
_NET_CURRENT_DESKTOP(CARDINAL) = 4294967295
_NET_NUMBER_OF_DESKTOPS(CARDINAL) = 0
[...]
```
These are set by `exwm-workspace--remove-frame-as-workspace` when exiting. Although I think openbox should make sure the values are sane before relying on them, we should also try not to set insane values.  Would it make sense for the mentioned function to ensure `_NET_NUMBER_OF_DESKTOPS` is never set to a value smaller than `1`, and that `_NET_CURRENT_DESKTOP` is always set to a value between 0 and `_NET_NUMBER_OF_DESKTOPS`?

ch11ng commented:
@medranocalvo Thanks! It was `_NET_NUMBER_OF_DESKTOPS`.

medranocalvo commented:
Thank you!
